### ISSUE TITLE
KRA installation with HSM: Unable to start KRA engine: Selftest failed: Certificate not found: NHSM-CONN-XC:transportCert cert-pki-kra

### ISSUE THREAD

**Issue by rcritten** [dogtagpki/pki#4335]
Using an nCipher HSM I can install IPA (with development patches) but installing a KRA fails because the transportCert cannot be found in the NSS database. It's there though.

```
certutil -L -d /etc/pki/pki-tomcat/alias -h all

Certificate Nickname                                         Trust Attributes
                                                             SSL,S/MIME,JAR/XPI

Enter Password or Pin for "NHSM-CONN-XC":
auditSigningCert cert-pki-kra                                u,u,Pu
storageCert cert-pki-kra                                     u,u,u
transportCert cert-pki-kra                                   u,u,u
auditSigningCert cert-pki-ca                                 u,u,Pu
subsystemCert cert-pki-ca                                    u,u,u
Server-Cert cert-pki-ca                                      u,u,u
ocspSigningCert cert-pki-ca                                  u,u,u
caSigningCert cert-pki-ca                                    CTu,Cu,Cu
NHSM-CONN-XC:caSigningCert cert-pki-ca                       CTu,Cu,Cu
NHSM-CONN-XC:ocspSigningCert cert-pki-ca                     u,u,u
NHSM-CONN-XC:subsystemCert cert-pki-ca                       u,u,u
NHSM-CONN-XC:auditSigningCert cert-pki-ca                    u,u,Pu
NHSM-CONN-XC:transportCert cert-pki-kra                      u,u,u
NHSM-CONN-XC:storageCert cert-pki-kra                        u,u,u
NHSM-CONN-XC:auditSigningCert cert-pki-kra                   u,u,Pu
```

Now this output makes it look like the keys are also in the default token but they aren't:

```
certutil -L -d /etc/pki/pki-tomcat/alias 

Certificate Nickname                                         Trust Attributes
                                                             SSL,S/MIME,JAR/XPI

caSigningCert cert-pki-ca                                    CT,C,C
ocspSigningCert cert-pki-ca                                  ,,   
Server-Cert cert-pki-ca                                      u,u,u
subsystemCert cert-pki-ca                                    ,,   
auditSigningCert cert-pki-ca                                 ,,P  
transportCert cert-pki-kra                                   ,,   
storageCert cert-pki-kra                                     ,,   
auditSigningCert cert-pki-kra                                ,,P  
```

Installation was done after an IPA install with: `
[kra.ini.txt](https://github.com/dogtagpki/pki/files/10787500/kra.ini.txt)
pkispawn -s KRA -f /root/kra.ini --debug 2>&1 | tee log`

Including the spawn and debug logs

[debug.2023-02-20.log.txt](https://github.com/dogtagpki/pki/files/10787494/debug.2023-02-20.log.txt)
[log.txt](https://github.com/dogtagpki/pki/files/10787495/log.txt)

**Comment by rcritten** [comment]
[kra.ini.txt](https://github.com/dogtagpki/pki/files/10787502/kra.ini.txt)

**Comment by edewata** [comment]
Is this an nCipher-specific problem? I suppose KRA installation works fine with SoftHSM.

**Comment by rcritten** [comment]
It seems so. It does work with softhsm.

**Comment by edewata** [comment]
@celestian @ckelleyRH

**Comment by HiTShiroya** [comment]
@rcritten I was facing the same issue while testing Key archival and key recovery for LunaHSM, in my case, I ran this script http://pastebin.test.redhat.com/1093445 and then reinstall the subsystem and it worked, could you please check this may it work in your case as well.

**Comment by rcritten** [comment]
Thanks for the suggestion. Same behavior, it can't find the token:

2023-03-02 11:54:31 [Catalina-utility-1] INFO: TransportKeyUnit: Loading NHSM-CONN-XC:transportCert cert-pki-kra certificate
2023-03-02 11:54:31 [Catalina-utility-1] WARNING: KeyRecoveryAuthority: transport unit exception Internal Error: org.mozilla.jss.crypto.ObjectNotFoundException: Certificate not found: NHSM-CONN-XC:transportCert cert-pki-kra
Internal Error: org.mozilla.jss.crypto.ObjectNotFoundException: Certificate not found: NHSM-CONN-XC:transportCert cert-pki-kra

It is definitely there and I can manipulate it with certutil.

**Comment by kimettog** [comment]
Hi Rob
I do not see your issue...when I use my ca.cfg and kra.cfg but I noticed your was missing these params among others?
pki_hsm_libfile = /opt/nfast/toolkits/pkcs11/libcknfast.so
pki_hsm_modulename = nfast
pki_client_pkcs12_password = SECret.123
pki_client_dir = /opt/topology-02-KRA
pki_ca_signing_cert_path=/etc/pki/topology-02-CA/alias/ca.crt
pki_cert_chain_path=/etc/pki/topology-02-CA/alias/ca.crt

Would you try with these 2 cfg files and tweak to (edit::will send cfg file to you directly )

**Comment by rcritten** [comment]
The pki_hsm_libfile and pki_hsm_modulename aren't needed if the token is already configured (this adds the module to the NSS database).

the other options, AFAIU, aren't needed. IPA has never used them.

Thanks for the suggestions. I'll keep poking at it.

**Comment by rcritten** [comment]
Digging through the selfsign log the KRA fails to start with:

Certificate NHSM-CONN-XC:auditSigningCert cert-pki-kra is invalid: Invalid certificate: (-8101) Certificate type not approved for application.

This may be another instance of "I can't get the certificate". I'm not sure.

**Comment by rcritten** [comment]
I've also seen it fail with:

0.Catalina-utility-2 - [21/Mar/2023:15:58:34 EDT] [20] [1] SystemCertsVerification: system certs verification failure: Certificate not found: NHSM-CONN-XC:transportCert cert-pki-kra

Though certutil confirms the cert by that nickname exists.

Restarting the service (pki-tomcatd@pki-tomcat) brings the KRA up.

I wonder if it's how the service is being started/restarted in base/server/python/pki/server/__init__.py::deploy_webapp

I don't know enough about tomcat, but is it supposed to see the new deployment file and load it up without requiring a service restart? I have the feeling this could be related to IPA using a single tomcat instance instead of two.

**Comment by rcritten** [comment]
The issue looks related to tomcat picking up the new xml configuration file and making the kra available. For some reason it isn't doing it.

I have a kludgy way to get a KRA install to complete in IPA by manually restarting pki-tomcat.

In fact, I need to restart all of IPA after that in order to avoid a permission error:

2023-03-22 08:39:42 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-8] INFO: PKIService: Response class: Account
2023-03-22 08:39:42 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-9] INFO: UGSubsystem: Retrieving user uid=ipakra,ou=People,o=kra,o=ipaca
2023-03-22 08:39:42 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-9] SEVERE: checkPermission(): permission denied for the resource certServer.kra.keys on operation execute
2023-03-22 08:39:42 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-9] SEVERE: AAclAuthz: Authorization Failed
certServer.kra.keys does not have permission to execute

Adding a vault fails with:

pki.PKIException: Unable to decrypt passphrase: Failed to unwrap key: (-8186) security library: invalid algorithm.

The IPA I'm testing is still using PKCS1v15. I tried OEAP as well and it failed with a similar algorithm error.

I don't know how to directly test vault commands from the command-line with a pure CA/KRA installation so I can't verify that it works without IPA.

**Comment by rcritten** [comment]
Can I get some help with the kra cli? I just want to do a simple vault creation and key retrieval. There seems to be limited cli documentation for this, the most I can find is the man page but while it has some decent usage examples it is thin on details and particularly use-cases (e.g. to store data to X, Y then Z).

**Comment by edewata** [comment]
Sorry, I'm not too familiar with KRA CLI for archival/recovery and unfortunately we do not have the capacity in the current sprint to do an investigation for this ticket, but QE might be able to explain how archival/recovery is tested.

**Comment by edewata** [comment]
If this is urgent please discuss the priority with @celestian.

**Comment by rcritten** [comment]
I've duplicated the problem without IPA. If deploying the CA and KRA on separate ports, with separate tomcat instances it works. When deploying them together it fails the same way as with IPA: tomcat isn't loading the kra deployment file.

A restart of the tomcat service while the KRA is checking to see if it is started will complete the installation process.

I can't confirm that the KRA actually works because in any situation (separate instances or together) I'm unable to store keys in the KRA. I followed https://github.com/dogtagpki/pki/wiki/PKI-KRA-Key-CLI closely and in all cases requests for key storage are created but no amount of approving will provide the keyId.
